File: server/src/main/java/by/brawl/ws/holder/gamesession/GameSessionsPool.java
package by.brawl.ws.holder.gamesession;

import by.brawl.util.Exceptions;
import by.brawl.util.Mappers;
import by.brawl.ws.dto.BattlefieldDto;
import by.brawl.ws.dto.MulliganDto;
import by.brawl.ws.holder.BattlefieldHolder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

@Component
public class GameSessionsPool {

	private static final Logger LOG = LoggerFactory.getLogger(GameSessionsPool.class);

	private Map<String, GameSession> gameSessions = new HashMap<>();

	public GameSession getSession(String principalName) {
		return gameSessions.get(principalName);
	}

	public void putSession(WebSocketSession session) {
		String key = session.getPrincipal().getName();
		GameSession value = new GameSession(session);
		gameSessions.put(key, value);
	}

	private Set<GameSession> getAllReceiversByBattlefield (BattlefieldHolder battlefield) {
        Set<GameSession> receivers = Mappers.asSet(battlefield.getConnectedAccountsIds(), key -> gameSessions.get(key));
        if (receivers.contains(null)) {
            throw Exceptions.produceNullPointer(LOG, "BattlefieldHolder contains null sessions!");
        }
		return receivers;
	}

	public void sendMulliganData(BattlefieldHolder battlefield) {
		getAllReceiversByBattlefield(battlefield).forEach(receiver -> {
			MulliganDto dto = new MulliganDto(
					receiver.getBattlefieldHolder().getMulliganHeroes(),
					receiver.getId(),
					receiver.getBattlefieldHolder().getGameState()
			);
			receiver.sendDto(dto);
		});
	}

	public void sendBattlefieldData(BattlefieldHolder battlefield) {
		getAllReceiversByBattlefield(battlefield).forEach(receiver -> {
			BattlefieldDto dto = new BattlefieldDto(receiver.getBattlefieldHolder(), receiver.getId());
			receiver.sendDto(dto);
		});
	}
}

File: server/src/main/java/by/brawl/ws/holder/gamesession/GameSessionsPool.kt
package by.brawl.ws.holder.gamesession

import by.brawl.ws.dto.BattlefieldDto
import by.brawl.ws.dto.HeroDto
import by.brawl.ws.dto.MulliganDto
import by.brawl.ws.holder.RoomHolder
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession
import java.util.*

@Component
open class GameSessionsPool {

    private val gameSessions = HashMap<String, GameSession>()

    fun getSession(principalName: String): GameSession {
        return gameSessions[principalName] ?: throw IllegalArgumentException("Session for $principalName doesn't exist")
    }

    fun putSession(session: WebSocketSession) {
        val key = session.principal.name
        val value = GameSession(session)
        gameSessions.put(key, value)
    }

    private fun getAllReceiversByBattlefield(room: RoomHolder): Set<GameSession> {
        return room.connectedAccountsIds.mapNotNull { key -> gameSessions[key] }.toSet()
    }

    fun sendMulliganData(room: RoomHolder) {
        getAllReceiversByBattlefield(room).forEach { receiver ->
            val mulliganHolder = receiver.roomHolder.mulliganHolder
            val dto = MulliganDto(
                    mulliganHolder.ownHeroes(receiver.username).map { HeroDto(it) },
                    mulliganHolder.enemyHeroes(receiver.username).map { HeroDto(it) },
                    receiver.roomHolder.gameState
            )
            receiver.sendDto(dto)
        }
    }

    fun sendBattlefieldData(room: RoomHolder) {
        getAllReceiversByBattlefield(room).forEach { receiver ->
            val dto = BattlefieldDto(receiver.roomHolder, receiver.username)
            receiver.sendDto(dto)
        }
    }
}